src/psyclone/tests/exceptions_test.py

```diff
@@ -0,0 +1,138 @@
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. J. Voysey, Met Office
+
+''' Test exception classes to ensure consistent __repr__ & __str__ methods. '''
+
+from __future__ import absolute_import
+from psyclone.errors import PsycloneError
+
+import sys
+import pkgutil
+import pytest
+import psyclone
+import inspect
+import os
+import importlib
+
+class DummyPsycloneError(PsycloneError):
+    ''' Provides a dummy PSyclone specific error class as for use in this test
+    '''
+    def __init__(self):
+        PsycloneError.__init__(self, "")
+        self.value = "Dummy Psyclone Error"
+
+
+def all_sub_exceptions(expt):
+    ''' Recursively find the set of all the exceptions which are subclasses of
+        the given exception class. '''
+
+    new_sub_except = [add_except for sub_except in expt.__subclasses__()
+                     for add_except in all_sub_exceptions(sub_except)]
+    return set(expt.__subclasses__()).union(new_sub_except)
+
+
+def import_submodules(package, recursive=True):
+    """ Import all submodules of a module, recursively, including subpackages
+
+    :param package: package (name or actual module)
+    :type package: str | module
+
+    :rtype: dict[str, types.ModuleType]
+    """
+    if isinstance(package, str):
+        package = importlib.import_module(package)
+    results = {}
+    for loader, name, is_pkg in pkgutil.walk_packages(package.__path__):
+        full_name = package.__name__ + '.' + name
+        if "test" not in full_name:
+            results[full_name] = importlib.import_module(full_name)
+            if recursive and is_pkg:
+                results.update(import_submodules(full_name))
+    return results
+
+
+def test_exception_repr():
+    ''' Test the properties of Exception classes defined by Psyclone. '''
+
+    modules = dict()
+
+    # Recursively walk through the psyclone module, importing sub-modules.
+    # Store any class definitions we come across.
+
+    modules = import_submodules("psyclone")
+    for mod in modules:
+        package = importlib.import_module(mod)
+
+    all_excpetions = all_sub_exceptions(Exception)
+    psy_excepts = [exc for exc in all_excpetions if "psyclone." in str(exc)]
```

Comment:
Could we now have `PSycloneError` instead of `Exception` on the line above and then remove this line entirely?

Reply:
In principle yes - but I think I would argue this should remain as it is to make the test more robust. If a future change added an exception that (accidentally) did not inherit `PSycloneError`, it would be still be picked up by the current formulation; however it would be missed if suggested change was made.

Reply:
Actually, could we therefore go one better and check that every exception we find is an instance of `PSycloneError`. Should just be one or two extra lines?